File: frontend/types/document.ts
// frontend/types/document.ts

// Document version properties
export interface DocumentVersionBase {
  version_number: number;
  content: string;
  created_by: string;
  notes?: string;
}

// Interface for document version creation
export interface DocumentVersionCreate extends DocumentVersionBase {}

// Full document version model
export interface DocumentVersion extends DocumentVersionBase {
  created_at: string;
}

// Base document properties
export interface DocumentBase {
  title: string;
  document_type: string;
  client_id: string;
  project_id?: string;
  status: string;
  requires_signature: boolean;
}

// Interface for document creation
export interface DocumentCreate extends DocumentBase {
  content: string; // Initial content for version 1
}

// Interface for document updates
export interface DocumentUpdate {
  title?: string;
  document_type?: string;
  client_id?: string;
  project_id?: string;
  status?: string;
  requires_signature?: boolean;
  new_version_content?: string;
  new_version_notes?: string;
}

// Full document model including server-generated fields
export interface Document extends DocumentBase {
  _id: string;
  current_version: number;
  versions: DocumentVersion[];
  signed: boolean;
  signed_at?: string;
  signed_by?: string;
  created_by: string;
  created_at: string;
  updated_at: string;
}

File: backend/app/schemas/document.py
# backend/app/schemas/document.py
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict


class DocumentVersionBase(BaseModel):
    version_number: int
    content: str
    created_by: str
    notes: Optional[str] = None


class DocumentVersionCreate(DocumentVersionBase):
    pass


class DocumentVersionResponse(DocumentVersionBase):
    created_at: datetime


class DocumentBase(BaseModel):
    title: str
    document_type: str
    client_id: str
    project_id: Optional[str] = None
    status: str = "draft"
    requires_signature: bool = False


class DocumentCreate(DocumentBase):
    content: str  # Initial content for version 1


class DocumentUpdate(BaseModel):
    title: Optional[str] = None
    document_type: Optional[str] = None
    client_id: Optional[str] = None
    project_id: Optional[str] = None
    status: Optional[str] = None
    requires_signature: Optional[bool] = None
    new_version_content: Optional[str] = None
    new_version_notes: Optional[str] = None


class DocumentResponse(DocumentBase):
    id: str = Field(..., alias="_id")
    current_version: int
    versions: List[DocumentVersionResponse]
    signed: bool
    signed_at: Optional[datetime] = None
    signed_by: Optional[str] = None
    created_by: str
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(
        populate_by_name=True,
    )


class DocumentInDB(DocumentResponse):
    pass